src/pykeen/nn/pyg.py

```diff
@@ -0,0 +1,391 @@
+    entity_initializer = LabelBasedInitializer.from_triples_factory(
+        triples_factory=dataset.training,
+        for_entities=True,
+    )
+    (embedding_dim,) = entity_initializer.tensor.shape[1:]
+    r = pipeline(
+        dataset=dataset,
+        model=ERModel,
+        model_kwargs=dict(
+            interaction="distmult",
+            entity_representations="UniRelationalMessagePassing",
+            entity_representations_kwargs=dict(
+                triples_factory=dataset.training,
+                base_kwargs=dict(
+                    shape=embedding_dim,
+                    initializer=entity_initializer,
+                    trainable=False,
+                ),
+                layers=["GCN"] * 2,
+                layers_kwargs=dict(in_channels=embedding_dim, out_channels=embedding_dim),
+            ),
+            relation_representations_kwargs=dict(
+                shape=embedding_dim,
+            ),
+        ),
+    )
+"""
+from abc import abstractmethod
+from typing import Optional, Sequence
+
+import torch
+from class_resolver import ClassResolver, HintOrType, OneOrManyHintOrType, OneOrManyOptionalKwargs, OptionalKwargs
+from class_resolver.contrib.torch import activation_resolver
+from torch import nn
+
+from .representation import Representation
+from ..triples.triples_factory import CoreTriplesFactory
+from ..typing import OneOrSequence
+from ..utils import upgrade_to_sequence
+
+__all__ = [
+    "UniRelationalMessagePassingRepresentation",
+    "FeaturizedRelationTypeMessagePassingRepresentation",
+    "CategoricalRelationTypeMessagePassingRepresentation",
+]
+
+try:
+    from torch_geometric.nn.conv import MessagePassing
+
+    layer_resolver = ClassResolver.from_subclasses(
+        base=MessagePassing,  # type: ignore
+        suffix="Conv",
+    )
+except ImportError:
+    MessagePassing = None
+    layer_resolver = None
+
+_PYG_INSTALLATION_TEXT = """
+Requires `torch_geometric` to be installed.
+
+Please refer to
+
+    https://pytorch-geometric.readthedocs.io/en/latest/notes/installation.html
+
+for installation instructions.
+"""
+
+
+class MessagePassingRepresentation(Representation):
+    """
+    An abstract representation class utilizing PyTorch Geometric message passing layers.
+
+    It comprises:
+        * base (entity) representations, which can also be passed as hints
+        * a sequence of message passing layers. They are utilized in an abstract
+          :meth:`MessagePassingRepresentation._message_passing` to enrich the base representations
+          by neighborhood information.
+        * a sequence of activation layers in between the message passing layers.
+        * an `edge_index` buffer, which stores the edge index and is moved to the device alongside
+          the module.
+    """
+
+    #: the message passing layers
+    layers: Sequence[MessagePassing]
+
+    #: the edge index, shape: (2, num_edges)
+    edge_index: torch.LongTensor
```

Comment:
I am a bit hesitant to bake the edge index right into the class instead of passing it as an argument to the message passing layer - could you please elaborate on the motivation? 

With the edge index baked in, I am not sure the same GNN architecture can be used for training and inference graphs in the inductive setup 🤔

Reply:
I kept it similar to what we have in (inductive) `NodePiece`, where we also would have [separate representations for training & inference graph](https://github.com/pykeen/pykeen/blob/ba3c72adc924c34d4ee3cb4609f2519009cf6f20/src/pykeen/models/inductive/inductive_nodepiece.py#L142-L155), and the representations themselves are specific to a certain triples factory / graph.

Reply:
I'd also rather get the PR done and solve the inductive problem later :)